File: dynamic-programming/使用最小花费爬楼梯/java/Solution_Dp.java
class Solution_Dp {
    public int minCostClimbingStairs(int[] cost) {
        int n = cost.length;
        if (n == 1) {
            return cost[0];
        }
        if (n == 2) {
            return Math.min(cost[0], cost[1]);
        }
        int[] dp = new int[n + 1];
        dp[0] = 0;
        dp[1] = 0;
        for (int i = 2; i < n + 1; i++) {
            dp[i] = Math.min(cost[i - 1] + dp[i - 1], cost[i - 2] + dp[i - 2]);
        }
        return dp[n];
    }
}

File: dynamic-programming/grid/minimum-path-sum/ts/solution_dp.ts
function minPathSum(grid: number[][]): number {
    const m = grid.length, n = grid[0].length
    const dp: number[][] = Array.from({length: m + 1}, (): number[] => Array(n + 1))
    for (let i = 0; i <= m; i++) {
        for (let j = 0; j <= n; j++) {
            if (i === 0 || j === 0) {
                dp[i][j] = Number.MAX_SAFE_INTEGER
                continue
            }
            if (i === 1 && j === 1) {
                dp[1][1] = grid[0][0]
                continue
            }
            dp[i][j] = Math.min(dp[i - 1][j], dp[i][j - 1]) + grid[i - 1][j - 1]
        }
    }
    return dp[m][n]
}

export { minPathSum }